src/agent.c

```diff
@@ -454,12 +454,12 @@ void agent_run(juice_agent_t *agent) {
 		}
 
 		if (FD_ISSET(agent->sock, &readfds)) {
-			mutex_lock(&agent->mutex);
+			mutex_unlock(&agent->mutex);
```

Comment:
The mutex should be locked when entering `agent_recv`. I think the issue is likely to be that another thread kept the lock when it shouldn't.

Reply:
Ok

Reply:
Sorry, I didn't explain in too much detail.

What it was happening in my case was that the agent_run thread was locking the mutex and checking for receiving data. But there was no data to get, and the agent_recv wasn't returning.
In the meantime, the agent_send method was called. When this method tried to lock the mutex, it would block until the agent_run thread unlocked it. But this was never happening because agent_recv was blocked.

Somehow, if I don't allow the agent_send method to run while the agent_recv is waiting for data, nothing is received.

Maybe agent_recv could lock/unlock inside on the specific parts where is not thread safe.

Reply:
Oh I see. Then this is an issue with `agent_recv`, as the function should return as soon as possible, and it should never wait for data. Does `recvfrom()` block? If so, on what OS do you run it?

Reply:
> Does recvfrom() block?
I didn't check yet where exactly inside agent_recv is blocking. I can do that a bit later.

> If so, on what OS do you run it?
I run it in two Windows 10 laptops. One fails always, and the other never.